app/src/main/java/net/aiscope/gdd_app/ui/mask/customview/MaskLayer.kt

```diff
@@ -5,106 +5,197 @@ import android.graphics.Bitmap
+    fun drawMove(x: Float, y: Float) {
+        val (latestX, latestY) = currentPath.latestPoint
+        val dX = abs(x - latestX)
+        val dY = abs(y - latestY)
+        val touchTolerance = ViewConfiguration.get(context).scaledTouchSlop
+        if (dX >= touchTolerance || dY >= touchTolerance) {
+            currentPath.quadTo(x, y)
+            if (pathsAndPaints.isEmpty() || currentPath != pathsAndPaints.last().first) {
+                addPath(currentPath)
+            }
+        }
     }
 
-    private fun recreateBrush() {
-        val scale = getScaleFromMatrix(this.scaleMatrix)
+    fun onSaveInstanceState() = CustomViewBaseState(pathsAndPaints, undoPendingPaths)
 
-        maskPaintBrush.apply {
-            strokeWidth = viewHeight.coerceAtLeast(viewWidth).toFloat() / (TWENTY * scale)
+    fun onRestoreInstanceState(savedState: Parcelable) {
+        if (savedState is CustomViewBaseState) {
+            undoPendingPaths = savedState.undoPendingPaths
+            pathsAndPaints.addAll(savedState.composePathsAndPaints())
         }
+    }
 
-        maskPaintErase.apply {
-            strokeWidth = viewHeight.coerceAtLeast(viewWidth).toFloat() / (TWENTY * scale)
+    private fun addPath(path: PointToPointPath) {
+        for (i in pathsAndPaints.size - 1 downTo pathsAndPaints.size - undoPendingPaths) {
+            pathsAndPaints.removeAt(i)
         }
+        undoPendingPaths = 0
+        pathsAndPaints.add(path to currentPaint)
     }
 
-    fun enterDrawMode() {
-        maskPaint = maskPaintBrush
-    }
+    private fun scaleBrush() {
+        val scale = getScaleFrom(this.scaleMatrix)
 
-    fun enterEraseMode() {
-        maskPaint = maskPaintErase
+        currentPaint = newDefaultPaintBrush(context.getColor(R.color.colorPrimary)).apply {
+            val (viewWidth, viewHeight) = viewDimensions
+            strokeWidth = viewHeight.coerceAtLeast(viewWidth).toFloat() / (TWENTY * scale)
+        }
     }
 
-    private fun getScaleFromMatrix(m: Matrix): Float {
+    private fun getScaleFrom(m: Matrix): Float {
         val p = FloatArray(MATRIX_SIZE)
         m.getValues(p)
 
         return p[Matrix.MSCALE_X]
     }
 
-    fun getBitmap(): Bitmap? = maskBitmap
+    class CustomViewBaseState(
+        pathsAndPaints: List<Pair<PointToPointPath, Paint>>,
+        val undoPendingPaths: Int
+    ) : Parcelable {
 
-    companion object {
-        private const val MATRIX_SIZE = 9
-        private const val MASK_PAINT_ALPHA = 200
+        companion object CREATOR : Parcelable.Creator<CustomViewBaseState> {
+            override fun createFromParcel(parcel: Parcel): CustomViewBaseState {
+                return CustomViewBaseState(parcel)
+            }
+
+            override fun newArray(size: Int): Array<CustomViewBaseState?> {
+                return arrayOfNulls(size)
+            }
+        }
+
+        val basePathsAndPaintChangesData: MutableList<Pair<PathBaseData, PaintChangeBaseData?>> = LinkedList()
```

Comment:
Is this public so just that it can be used in the tests? I would suggest moving the logic to a collaborator which also could be more unit-testable, or at least snap a `@VisibleForTesting(otherwise = PRIVATE)` annotation to this field :D